### PR TITLE
[Hub Generated] Review request for Microsoft.DocumentDB to add version stable/2023-03-15

### PR DESCRIPTION
**Pull request by ghost**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/Azure/azure-rest-api-specs/carjackson-msft-cosmos-db-Microsoft.DocumentDB-2023-03-15...main/).
## ARM API Information (Control Plane)

Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: custom-words.txt [hunk 1]
```diff
@@ -319,6 +319,7 @@ BRCPF
 browsable
 bruceper
 Bsaling
+buildout
```
Comment (TimLovellSmith) [inline custom-words.txt:322]: nit.. is buildout a real word?

**Comment by TimLovellSmith** [comment]
>         },

ARM recommends using enums over booleans for new properties, to ensure future proof APIs.

Standard guidance is: replace boolean/switch properties with better enum whenever possible.
A boolean will forever have two valid values (true or false). A string enum is more flexible and future proof option because it allows additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].. Also, properties should always ideally provide more meaningful values just than True and False, which are meaningless without context. For example two boolean switches "isTypeA" and "isTypeB" can be replaced with one enum "type": [A, B, DefaultType]. 
Note: do NOT just define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its still confusing, thanks to the boolean value names. Hint: Try to think of enum value names that look meaningful even without context like 'Unencrypted', 'Encrypted', 'TlsEncrypted' which you can use instead.

[Conversely: do not do breaking changes by change boolean properties in use by customers with existing API versions.]

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9830 in 9f8ff02. [](commit_id = 9f8ff02adf9db7388e6a0760c35feed70dc43edf, deletion_comment = False)

**Comment by TimLovellSmith** [comment]
>           "description": "Current buildout status of the region."

What would the possible values of this property mean?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9834 in 9f8ff02. [](commit_id = 9f8ff02adf9db7388e6a0760c35feed70dc43edf, deletion_comment = False)

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/examples/CosmosDBDatabaseAccountListConnectionStrings.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  "responses": {
+    "200": {
+      "body": {
+        "connectionStrings": [
+          {
+            "connectionString": "connection-string",
+            "description": "Name of the connection string",
+            "keyKind": "Kind of the connection string key",
+            "type": "Type of the connection string"
```
Comment (TimLovellSmith) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/examples/CosmosDBDatabaseAccountListConnectionStrings.json:16]: could these be more realistic examples?

File: custom-words.txt (hunk 1 above)
Comment (udit3333) [reply custom-words.txt:322]: `buildout` here refers to the status of Azure Cosmos DB in a region. For new regions which are still in buildout the status can be `Initializing` based on the stage of regional buildout.

**Comment by udit3333** [comment]
> > ```
> >       "description": "Current buildout status of the region."
> > ```
> 
> What are the possible values of this property? What would they mean?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9834 in [5bcf280](https://github.com/Azure/azure-rest-api-specs/commit/5bcf2804a0b5788fed83013877e8639426718d9e). [](commit_id = [5bcf280](https://github.com/Azure/azure-rest-api-specs/commit/5bcf2804a0b5788fed83013877e8639426718d9e), deletion_comment = False)

`status` here refers to the buildout status of Azure Cosmos DB in a region. For new regions which are still in buildout the status can be `Uninitialized`, `Initializing`, `InternallyReady`, `Online`, `Deleting` based on the stage of regional buildout.

**Comment by udit3333** [comment]
> > ```
> >     },
> > ```
> 
> ARM recommends using enums over booleans for new properties, to ensure future proof APIs.
> 
> Standard guidance is: replace boolean/switch properties with better enum whenever possible. A boolean will forever have two valid values (true or false). A string enum is more flexible and future proof option because it allows additional values to be added in the future in a non-breaking way, e.g. [Enabled, Disabled, Suspended, Deallocated].. Also, properties should always ideally provide more meaningful values just than True and False, which are meaningless without context. For example two boolean switches "isTypeA" and "isTypeB" can be replaced with one enum "type": [A, B, DefaultType]. Note: do NOT just define a 'boolean enum' with two values 'True and False'. This might be easier to 'extend' in terms of types, but semantically its still confusing, thanks to the boolean value names. Hint: Try to think of enum value names that look meaningful even without context like 'Unencrypted', 'Encrypted', 'TlsEncrypted', or 'Allowed', 'NotAllowed' which you can use instead.
> 
> [Conversely: do not do breaking changes by change boolean properties in use by customers with existing API versions.]
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9830 in [5bcf280](https://github.com/Azure/azure-rest-api-specs/commit/5bcf2804a0b5788fed83013877e8639426718d9e). [](commit_id = [5bcf280](https://github.com/Azure/azure-rest-api-specs/commit/5bcf2804a0b5788fed83013877e8639426718d9e), deletion_comment = False)

For properties `isSubscriptionRegionAccessAllowedForRegular` and `isSubscriptionRegionAccessAllowedForAz` there are only two possible values either the subscription will have access or not. That's why we have kept as boolean. 
`status` we can update to use enum since it can multiple values as mentioned above.  Let me know what you think?

**Comment by udit3333** [comment]
> > ```
> >     "isSubscriptionRegionAccessAllowedForRegular": {
> > ```
> 
> What does Regular mean? Is ForRegular worth having in the name?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9821 in [9f8ff02](https://github.com/Azure/azure-rest-api-specs/commit/9f8ff02adf9db7388e6a0760c35feed70dc43edf). [](commit_id = [9f8ff02](https://github.com/Azure/azure-rest-api-specs/commit/9f8ff02adf9db7388e6a0760c35feed70dc43edf), deletion_comment = False)

This is the terminology we used for non zonal region restrictions. Since Cosmos DB can have separate level of region restrictions for both zonal and non-zonal in a region. This is used to avoid confusion between the both.

**Comment by visingla-ms** [comment]
>         },

Another way of thinking converting these to enums is:

subscriptionRegionAccessState as enum with values ["NonAvailabilityZonesOnly", "AvailabilityZonesOnly", "All"]

Something like this. What do you think?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9830 in 3fc111d. [](commit_id = 3fc111dd8ad541fbf8404e11dec6cdfe920136c3, deletion_comment = False)

**Comment by visingla-ms** [comment]
>         "status": {

If these are known values, then definitely make it enum.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2023-03-15/cosmos-db.json:9831 in 3fc111d. [](commit_id = 3fc111dd8ad541fbf8404e11dec6cdfe920136c3, deletion_comment = False)

**Comment by visingla-ms** [comment]
A couple of suggestions about using enums. Let me know what do you think? I would suggest converting these to enums if possible.

**Comment by udit3333** [comment]
> @udit3333 A couple of suggestions about using enums. Let me know what do you think? I would suggest converting these to enums if possible.

@visingla-ms Updated the status property to enum. For `isSubscriptionRegionAccessAllowedForRegular` and `isSubscriptionRegionAccessAllowedForAz` it will require to refactor the RP handler and release to roll out. Keeping them as boolean instead.

**Comment by pjohari-ms** [comment]
@lirenhe We got an arm signoff. The Avocado test that's failing is due to APIs present in a much older version (2016) and we would like to get an exception for this.
Additionally, what are the ways we can fix this avocado exception going forward, considering that we don't have these APIs anymore. Should we update the older API version as a last resort?

**Comment by pjohari-ms** [comment]
@lirenhe Can you please merge the PR?

**Review by lirenhe — APPROVED** [review]

**Comment by pjohari-ms** [comment]
@lirenhe Please merge the PR. Thanks.

**Pull request merged**